File: leetcode/46-permutations/main.go
package main

import "fmt"

// recursive dfs
// time		O(n!)
// space	O(n!) due to the call stack
// beats 77.21%
func permute(nums []int) [][]int {
	result := [][]int{}
	var dfs func(arr []int, path []int)
	dfs = func(arr []int, path []int) {
		if len(arr) == 0 {
			result = append(result, path)
		} else {
			for i := 0; i < len(arr); i++ {
				temp := arr[i]

				copy := []int{}
				copy = append(copy, arr...)
				trim := append(copy[:i], copy[i+1:]...)

				nextpath := []int{}
				nextpath = append(nextpath, path...)
				nextpath = append(nextpath, temp)

				dfs(trim, nextpath)
			}
		}
	}
	dfs(nums, []int{})
	return result
}

// iterative insertion
//   1
//  ^ ^
//  2 2
//
//   2,1         1,2
//  ^ ^ ^       ^ ^ ^
//  3 3 3       3 3 3
// time		O(n!)
// space	O(1)
// beats 77.14%
func permute1(nums []int) [][]int {
	perms := [][]int{[]int{}}
	for i := 0; i < len(nums); i++ {
		new_perms := [][]int{}
		for j := 0; j < len(perms); j++ {
			perm := perms[j]
			for k := 0; k < len(perm)+1; k++ {

				// i really hate golang slice and concate
				cop := []int{}
				cop = append(cop, perm[:k]...)
				cop = append(cop, nums[i])
				cop = append(cop, perm[k:]...)

				new_perms = append(new_perms, cop)
			}
		}
		perms = new_perms
	}
	return perms
}

func main() {
	fmt.Println(permute([]int{1, 2, 3}))
	fmt.Println(permute1([]int{1, 2, 3}))
}

File: leetcode/46-permutations/main.py
"""
    1st: recursive dfs

    e.g. [1,2,3]
    1          2         3
    2,3 3,2   1,3  3,1    1,2 2,1

    select k from n number, it is a permutation problem
    therefore, the time complexity is nPk = n!/(n-k)!

    therefore, the crux is dfs(nums[:i]+arr[i+1:], prefix+[nums[i]])
    beats 32.64%
"""


class Solution(object):

    def __init__(self):
        self.result = []

    def permute(self, nums):
        self.dfs(nums, [])
        return self.result

    def dfs(self, cands, chosen):
        if len(cands) == 0:
            self.result.append(chosen)
        for i in range(len(cands)):
            self.dfs(cands[:i] + cands[i+1:], chosen + [cands[i]])


# s = Solution()
# print(s.permute([1, 2, 3]))

# iterative insertion
# e.g. [1,2,3]
#   1
#  ^ ^
#  2 2
#
#   2,1         1,2
#  ^ ^ ^       ^ ^ ^
#  3 3 3       3 3 3
# beats 32.64%


def permute1(nums):
    perms = [[]]
    for num in nums:  # for each number
        new_perms = []
        for perm in perms:  # for each temporary result
            for i in range(len(perm)+1):  # for each slot
                new_perms.append(perm[:i] + [num] + perm[i:])
        perms = new_perms
    return perms


print(permute1([1, 2, 3]))
